src/DSN.php

```diff
@@ -0,0 +1,241 @@
+ * @author Aaron Scherer <aequasi@gmail.com>
+ */
+class DSN
+{
+    /**
+     * @type string
+     */
+    protected $dsn;
+
+    /**
+     * @type string
+     */
+    protected $protocol;
+
+    /**
+     * @type string
+     */
+    protected $password;
+
+    /**
+     * @type string
+     */
+    protected $host;
+
+    /**
+     * @type int
+     */
+    protected $port;
+
+    /**
+     * @type string
+     */
+    protected $socket;
+
+    /**
+     * @type int
+     */
+    protected $database;
+
+    /**
+     * @type int
+     */
+    protected $weight;
+
+    /**
+     * @type string
+     */
+    protected $alias;
+
+    /**
+     * Constructor.
+     *
+     * @param string $dsn
+     */
+    public function __construct($dsn)
+    {
+        $this->dsn = $dsn;
+        $this->parseDsn($dsn);
+    }
+
+    /**
+     * @return string
+     */
+    public function getProtocol()
+    {
+        return $this->protocol;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getDatabase()
+    {
+        return $this->database;
+    }
+
+    /**
+     * @return int
+     */
+    public function getWeight()
+    {
+        return $this->weight;
+    }
+
+    /**
+     * @return string
+     */
+    public function getHost()
+    {
+        return $this->host;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPassword()
+    {
+        return $this->password;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getPort()
+    {
+        if (null !== $this->socket) {
+            return;
+        }
+
+        return $this->port ?: 6379;
+    }
+
+    /**
+     * @return string
+     */
+    public function getSocket()
+    {
+        return $this->socket;
+    }
+
+    /**
+     * @return string
+     */
+    public function getAlias()
+    {
+        return $this->alias;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isValid()
+    {
+        if (null === $this->getProtocol()) {
+            return false;
+        }
+
+        if (null !== $this->getHost() && null !== $this->getPort()) {
+            return true;
+        }
+
+        if (null !== $this->getSocket()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private function parseProtocol($dsn)
+    {
+        $regex = '/^(redis|mysql|mongodb|tcp):\/\//i';
+
+        preg_match($regex, $dsn, $matches);
+
+        if (isset($matches[1])) {
+            $this->protocol = $matches[1];
+        }
+    }
+
+    /**
+     * @param string $dsn
+     */
+    private function parseDsn($dsn)
+    {
+        $this->parseProtocol($dsn);
+
+        $dsn = str_replace($this->protocol.'://', '', $dsn); // remove "redis://"
+        if (false !== $pos = strrpos($dsn, '@')) {
+            // parse password
+            $this->password = str_replace('\@', '@', substr($dsn, 0, $pos));
+            $dsn            = substr($dsn, $pos + 1);
+        }
+        $dsn = preg_replace_callback(
```

Comment:
After this line there is too much regex for me. =)

It looks good. But does it work?